pilot/adapter/config/file/monitor.go

```diff
@@ -0,0 +1,361 @@
+	}
+
+	// Save the updated list.
+	m.resources = newResources
+}
+
+func (m *Monitor) createConfig(r *resource) {
+	cfg, err := r.toConfig()
+	if err != nil {
+		glog.Warningf("Failed to parse config (%m): %v", r.Metadata, err)
+		return
+	}
+
+	if _, err = m.store.Create(*cfg); err != nil {
+		glog.Warningf("Failed to create config (%m): %v ", r.Metadata, err)
+	}
+}
+
+func (m *Monitor) updateConfig(r *resource) {
+	cfg, err := r.toConfig()
+	if err != nil {
+		glog.Warningf("Failed to parse config (%m): %v", r.Metadata, err)
+		return
+	}
+
+	// Set the resource version based on the existing resource.
+	if prev, exists := m.store.Get(cfg.Type, cfg.Name, cfg.Namespace); exists {
+		cfg.ResourceVersion = prev.ResourceVersion
+	}
+
+	if _, err = m.store.Update(*cfg); err != nil {
+		glog.Warningf("Failed to update config (%m): %v ", r.Metadata, err)
+	}
+}
+
+func (m *Monitor) deleteConfig(r *resource) {
+	schema, ok := schemaForKind(r.Kind)
+	if !ok {
+		glog.Warningf("Missing schema for %q ", r.Kind)
+		return
+	}
+
+	if err := m.store.Delete(schema.Type, r.Metadata.Name, r.Metadata.Namespace); err != nil {
+		glog.Warningf("Failed to delete config (%m): %v ", r.Metadata, err)
+	}
+}
+
+func (m *Monitor) readFiles() byKey {
+	var result []*resource
+
+	err := filepath.Walk(m.root, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		} else if !supportedExtensions[filepath.Ext(path)] || (info.Mode()&os.ModeType) != 0 {
+			return nil
+		}
+		data, err := ioutil.ReadFile(path)
+		if err != nil {
+			glog.Warningf("Failed to read %s: %v", path, err)
+			return err
+		}
+		for _, r := range parseYaml(data) {
+			if !m.kinds[r.Kind] {
+				// Ignore unsupported kinds
+				continue
+			}
+			result = append(result, r)
+		}
+		return nil
+	})
+	if err != nil {
+		glog.Warningf("failure during filepath.Walk: %v", err)
+	}
+
+	// Sort by the resource IDs.
+	sort.Sort(byKey(result))
+	return result
+}
+
+// ParseYamlFile is a utility method that parses the given yaml file and returns all of the Istio configs. This is
+// equivalent to calling ioutil.ReadPath, followed by ParseYaml.
+func ParseYamlFile(path string) ([]*model.Config, error) {
+	data, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+	return ParseYaml(data)
+}
+
+// ParseYaml is a utility method that parses the contents of a given yaml file and returns all of the Istio configs.
+func ParseYaml(data []byte) ([]*model.Config, error) {
+	resources := parseYaml(data)
+	if len(resources) == 0 {
+		return nil, nil
+	}
+
+	configs := make([]*model.Config, len(resources))
+	for i, r := range resources {
+		var err error
+		if configs[i], err = r.toConfig(); err != nil {
+			return nil, err
+		}
+	}
+	return configs, nil
+}
+
+// parseYaml parses the data for an entire yaml file and returns slices for individual byKey.
+func parseYaml(data []byte) []*resource {
+	chunks := bytes.Split(data, []byte("\n---\n"))
```

Comment:
Does this work with e.g. a leading `---` at the top of a config file? AFAICT there were no tests with multiple config documents in the same file - could you add a few cases to exercise this handling (e.g. `---` at top of file, two sets of `---` next to eachother with no content between, trailing `---`, etc).

Reply:
We use `k8s.io/apimachinery/pkg/util/yaml` elsewhere for multi-doc YAML parsing. See https://github.com/istio/istio/blob/f6c32121d96e90e907be2f63990680576612fc96/pilot/adapter/config/crd/conversion.go#L116 for an example.

Reply:
+1 ... done.